### ISSUE TITLE
False Negative: Unused parameters in annotated, Struts-specific, and override-related contexts are not fully recognized.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6542]
# False Negative: Unused parameters in annotated, Struts-specific, and override-related contexts are not fully recognized.

Version
7.22.0

## Checker
- Checker id: `category/java/bestpractices.xml/UnusedFormalParameter`
- Checker description: Detects unused formal parameters in private methods/constructors (or all methods/constructors if configured) unless the parameter name indicates it is intentionally unused.

## Description of the false negative
`diff/output` marks 11 false negatives for `category/java/bestpractices.xml/UnusedFormalParameter`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are suppression or annotations.

### `PosCase1.java`
Source: `output/codeql/UselessParameter/src/main/java/scensct/core/pos/PosCase1.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Non-override method with non-empty body and an unused parameter should be flagged as dead code.

```java
// Non-override method with non-empty body and an unused parameter should be flagged as dead code.
package scensct.core.pos;

public class PosCase1 {
    // Non-override method with non-empty body (contains a statement) and an unused parameter.
    public void process(String usedParam, int unusedParam) { // [REPORTED LINE]
        System.out.println(usedParam); // usedParam is accessed, unusedParam is never accessed.
    }
}
```

### `PosCase2.java`
Source: `output/codeql/UselessParameter/src/main/java/scensct/core/pos/PosCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Non-override method with empty body and an unused parameter, while an overriding method with non-empty body exists, should be flagged as dead code.

```java
// Non-override method with empty body and an unused parameter, while an overriding method with non-empty body exists, should be flagged as dead code.
package scensct.core.pos;

public class PosCase2 {
    // Base class with a method that will be overridden.
    static class Base {
        public void action(String param) {
            // Non-empty body in base, but this is not an override method.
            System.out.println("Base action");
        }
    }
    
    // Derived class that overrides the method with a non-empty body.
    static class Derived extends Base {
        @Override
        public void action(String param) {
            System.out.println(param); // Overriding method with non-empty body, param is used.
        }
    }
    
    // Non-override method with empty body and an unused parameter.
    public void helper(String unusedParam) { // [REPORTED LINE]
        // Empty body, parameter never accessed.
    }
}
```

### `PosCase1.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase1.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Non-private instance method in final class with @SuppressWarnings("rawtypes") and unused parameter should be flagged as unused.

```java
// Non-private instance method in final class with @SuppressWarnings("rawtypes") and unused parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;

public final class PosCase1 {
    @SuppressWarnings("rawtypes")
    public void methodWithUnusedParam(String used, Integer unusedParam) { // [REPORTED LINE]
        // Parameter 'used' is used, 'unusedParam' is never used.
        System.out.println(used);
    }
}
```

### `PosCase2.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Private instance method in final class with @SuppressWarnings("unchecked"), undocumented unused parameter should be flagged as unused.

```java
// Private instance method in final class with @SuppressWarnings("unchecked"), undocumented unused parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;

public final class PosCase2 {
    @SuppressWarnings("unchecked")
    private void privateMethod(String used, Integer unusedParam) { // [REPORTED LINE]
        // No javadoc for method, so parameter 'unusedParam' is undocumented.
        System.out.println(used);
    }
}
```

### `PosCase3.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase3.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Private instance method in final Struts Action subclass with @SuppressWarnings("rawtypes") and unused non-exempt parameter should be flagged as unused.

```java
// Private instance method in final Struts Action subclass with @SuppressWarnings("rawtypes") and unused non-exempt parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;
import org.apache.struts.action.Action;

public final class PosCase3 extends Action {
    @SuppressWarnings("rawtypes")
    private void strutsMethod(String used, Integer unusedParam) { // [REPORTED LINE]
        // Parameter type Integer is not an exempt Struts/Servlet type.
        System.out.println(used);
    }
}
```

### `PosCase4.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase4.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Private instance method in final Struts Action subclass with @SuppressWarnings("unchecked"), undocumented unused non-exempt parameter should be flagged as unused.

```java
// Private instance method in final Struts Action subclass with @SuppressWarnings("unchecked"), undocumented unused non-exempt parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;
import org.apache.struts.action.Action;

public final class PosCase4 extends Action {
    @SuppressWarnings("unchecked")
    private void method(String used, Integer unusedParam) { // [REPORTED LINE]
        // No javadoc, parameter 'unusedParam' is undocumented and non-exempt.
        System.out.println(used);
    }
}
```

### `PosCase5.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase5.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Non-private instance method in non-final class with @SuppressWarnings("rawtypes"), undocumented unused parameter should be flagged as unused.

```java
// Non-private instance method in non-final class with @SuppressWarnings("rawtypes"), undocumented unused parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;

public class PosCase5 {
    @SuppressWarnings("rawtypes")
    public void method(String used, Integer unusedParam) { // [REPORTED LINE]
        // No javadoc, parameter 'unusedParam' is undocumented.
        System.out.println(used);
    }
}
```

### `PosCase6.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase6.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Private instance method in non-final class with @SuppressWarnings("unchecked") and unused parameter should be flagged as unused.

```java
// Private instance method in non-final class with @SuppressWarnings("unchecked") and unused parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;

public class PosCase6 {
    @SuppressWarnings("unchecked")
    private void privateMethod(String used, Integer unusedParam) { // [REPORTED LINE]
        System.out.println(used);
    }
}
```

### `PosCase7.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase7.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Non-private instance method in non-final Struts Action subclass with @SuppressWarnings("rawtypes") and unused non-exempt parameter should be flagged as unused.

```java
// Non-private instance method in non-final Struts Action subclass with @SuppressWarnings("rawtypes") and unused non-exempt parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;
import org.apache.struts.action.Action;

public class PosCase7 extends Action {
    @SuppressWarnings("rawtypes")
    public void method(String used, Integer unusedParam) { // [REPORTED LINE]
        // Parameter type Integer is not an exempt Struts/Servlet type.
        System.out.println(used);
    }
}
```

### `PosCase8.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase8.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Non-private instance method in non-final Struts Action subclass with @SuppressWarnings("unchecked"), undocumented unused non-exempt parameter should be flagged as unused.

```java
// Non-private instance method in non-final Struts Action subclass with @SuppressWarnings("unchecked"), undocumented unused non-exempt parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;
import org.apache.struts.action.Action;

public class PosCase8 extends Action {
    @SuppressWarnings("unchecked")
    public void method(String used, Integer unusedParam) { // [REPORTED LINE]
        // No javadoc, parameter 'unusedParam' is undocumented and non-exempt.
        System.out.println(used);
    }
}
```

### `PosCase9.java`
Source: `output/sonarqube/UnusedMethodParameterCheck/src/main/java/scensct/core/pos/PosCase9.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Private instance method in non-final Struts Action subclass with @SuppressWarnings("rawtypes"), undocumented unused non-exempt parameter should be flagged as unused.

```java
// Private instance method in non-final Struts Action subclass with @SuppressWarnings("rawtypes"), undocumented unused non-exempt parameter should be flagged as unused.
package scensct.core.pos;

import java.lang.SuppressWarnings;
import org.apache.struts.action.Action;

public class PosCase9 extends Action {
    @SuppressWarnings("rawtypes")
    private void method(String used, Integer unusedParam) { // [REPORTED LINE]
        // No javadoc, parameter 'unusedParam' is undocumented and non-exempt.
        System.out.println(used);
    }
}
```

## Cause analysis
The rule is documented as follows: Detects unused formal parameters in private methods/constructors (or all methods/constructors if configured) unless the parameter name indicates it is intentionally unused.

In this set the problem shows up around suppression or annotations. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `category/java/bestpractices.xml/UnusedFormalParameter`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.

**Comment by oowekyala** [comment]
In all those cases the parameters are not reported because their name start with `unused`, which is expected and documented.
